### ISSUE TITLE
Add better presentation for matrices

### ISSUE THREAD

**Issue by lckr** [lckr/jupyterlab-variableInspector#3]
Add a matrix view similar to RStudio. For that, use the datagrid library from phosphorjs which is already put to use in the .csv viewer in jlab.

The variable inspector should show the first few entries of a matrix in the overview table and switch to a datagrid for a detailed presentation.

To avoid performance issues the whole dataframes not loaded in advance but only of the user requests the detailed matrix presentation.

**Comment by lckr** [comment]
#18 limits the queried matrices to a fixed number of rows.
TODO:
- Allow changing the limit.
- To allow the inspection of the whole matrix I see two possible paths: Either load the matrix in chunks on inspection request and store it until the matrix window is closed (memory intensive ) OR lazily load the remaining rows when viewed in the data grid viewer and delete rows outside the window (multiple requests, no updates if the kernel is busy)

**Comment by chappers** [comment]
@lckr this might get tricky - we will probably have to have two modes of operation:

1.  For "native" Python objects, being Pandas/Numpy
2.  Non-native Python objects, being Spark DataFrames (spark dataframes are called from JVM via py4j) or TensorFlow variables (tensorflow does not run native on python)

**For item 2)** I'm going on a limb and say for now, lets not bother supporting it at this present time - Spark makes this nearly impossible anyway as you can't really "slice" dataframes in the same way as Pandas

**For item 1)**, my own personal opinion is that I don't really care too much, so long as it "works". I would be leaning on the first approach in some kind of "streaming" or "async" way.

In my mind, loading in chunks would be something like .... perhaps there's a way to not have it "store" in memory by dumping the result to disc/streaming it out so that it isn't the kernel holding it (not sure if what I'm saying makes sense/even 100% possible in the current form)

Basically something like (not at all suggesting we write stuff to disc, but for purposes of example/pseudocode):

```py
response = {"schema": pd.io.json.build_table_schema(df), "data": df.to_dict(orient="records")}
reponse.do_some_stuff_so_kernel_can_throw_me_away("<some reference>")
return <some reference>
```

In handler.ts/performMatrixInspection somehow reads data based on `<some reference>`....

This is how it would work in my head - I don't have any idea how I would actually implement it yet or if its even sensible!

**Comment by pcd123** [comment]
Hi,
Thanks for working on this nice project. I dream of a good varaible explorer 'a la matlab'
Do you know the qgrid widget? It could be great to use it to inspect dataframes.
https://github.com/quantopian/qgrid

**Comment by lckr** [comment]
Hi @pcd123, 
Thanks for your suggestion.  I took a look at the qgrid project which is quite neat. On the first glance I'm not sure if we can use it as it heavily makes use of python libraries. We want this extension to be as language independent as possible since not every user might have ipython installed.
Nevertheless, I will have a closer look at the qgrids project tomorrow.
